File: LC/[1345] Jump Game IV - v2.java
class Solution {
    public int minJumps(int[] arr) {
        int N = arr.length;
        if(N==1) return 0; //if only one step. no need to jump

        //create a cache to record value to index (given a value, what are the indexes we can jump to)
        Map<Integer, List<Integer>> val2idx = new HashMap<>();
        for(int ii=0; ii<N; ++ii) val2idx.computeIfAbsent(arr[ii], xx -> new ArrayList<>()).add(ii);

        //declare/define a queue (bfs)
        Queue<Integer> qq = new ArrayDeque<>();//holding nodes we can reach in this jump

        //declare/define visited to record visited nodes for pruning
        Set<Integer> visited = new HashSet<>();

        //initialize queue with first node
        visited.add(0);
        qq.offer(0);

        //perform BFS
        int level = 0;
        while(!qq.isEmpty()){
            level+=1;//we already check when N=1 (no jump)
            int sz = qq.size();
            for(int ii=0; ii<sz; ++ii){//consume all nodes in this level
                var curr_idx = qq.poll();

                //prev
                var prev_idx = curr_idx-1;
                if(prev_idx >= 0 && !visited.contains(prev_idx)){
                    visited.add(prev_idx);
                    qq.offer(prev_idx);
                }

                //next
                var next_idx = curr_idx+1;
                if(next_idx < N && !visited.contains(next_idx)){
                    visited.add(next_idx);
                    qq.offer(next_idx);
                }
                //by value
                if(val2idx.containsKey(arr[curr_idx])){
                    for(var xx: val2idx.get(arr[curr_idx])){
                        if(visited.contains(xx)) continue;
                        visited.add(xx);
                        qq.offer(xx);
                    }
                    val2idx.remove(arr[curr_idx]); //already visited all reachible nodes. prune to avoid unnecessary loop
                }
            }
            if(visited.contains(N-1)) return level;
        }
        return -1;
    }
}

// * N=10^4
// * NN=5N
// * 5NLogN = 5*10^6
// * we cannot use DP -> if we later reach an index with same value we can jump backward, that would change the prev optimal solution
// * min jump -> shortest distance -> BFS

File: LC/[1345] Jump Game IV - v2.py
class Solution:
    def minJumps(self, arr: List[int]) -> int:
        """
        * N=5*10^4 -> 5*10^4*(2*10) -> 10^6
        """
        N = len(arr)
        # preprocess val2idx
        val2idx=defaultdict(list)
        for ii, nn in enumerate(arr):
            val2idx[nn].append(ii)
        
        # prepare BFS (shortest distance/min jump requred to reach end)
        qq = deque()
        level=0
        visited=set() 

        # initialize root node
        qq.append(0)
        visited.add(0)

        while (sz:=len(qq)) > 0:
            if N-1 in visited: return level
            for _ in range(sz): #for the level   
                curr_idx = qq.popleft()
                curr_val = arr[curr_idx]
                prev_idx = curr_idx-1
                next_idx = curr_idx+1

                # consider prev
                if prev_idx >=0 and prev_idx not in visited:
                    visited.add(prev_idx)
                    qq.append(prev_idx)
                
                # consider next
                if next_idx < N and next_idx not in visited:
                    visited.add(next_idx)
                    qq.append(next_idx)
                
                # consider idx with same values
                if curr_val in val2idx:
                    for idx in val2idx[curr_val]:
                        if idx in visited: continue
                        visited.add(idx)
                        qq.append(idx)
                    
                    del val2idx[curr_val]

            level+=1

        return -1 # if not reaching end